File: array_list.rb
class ArrayList
  attr_reader :size

  def initialize
    @array = FixedArray.new(10)
    @size = 0
  end

  def [](index)
    check_bounds(index)
    @array[index]
  end

  def []=(index, value)
    check_lower_bound(index)
    raise FillMeInError
    # This method should set the value at the given index such that
    # array_list[index] will return the value.
    #
    # If the value is bigger than the current size of the array, the array
    # should be replaced with a bigger array to fit the new index, and all
    # indexes between the former last element and the new index should be
    # initialized with nil.
    #
    # The size after this method is called depends on the index provided. An
    # existing index would not affect the size, but a value greater than the
    # last index will add the difference to the size.
  end

  def delete(index)
    check_bounds(index)
    raise FillMeInError
    # This method should delete the value at the provided index and return
    # it. The size should be 1 less than it was before this method was called.
  end

  def <<(value)
    raise FillMeInError
    # This method should append a value to the end of this ArrayList and
    # increase the size by 1. The return value must be self.
    self
  end

  def >>(value)
    raise FillMeInError
    # This method should prepend a value to the beginning of this ArrayList and
    # increase the size by 1. The return value must be self.
    self
  end

  private

  def check_bounds(index)
    check_lower_bound(index)
    check_upper_bound(index)
  end

  def check_lower_bound(index)
    raise IndexError, "Invalid index: #{index}" if index < 0
  end

  def check_upper_bound(index)
    raise IndexError, "Invalid index: #{index}" if index >= size
  end
end

File: java/src/main/java/com/on_site/kamayan/collections/ArrayList.java
package com.on_site.kamayan.collections;

import com.on_site.kamayan.Kamayan;

public class ArrayList {
    private Object[] array;
    private int size;

    public ArrayList() {
        array = new Object[10];
        size = 0;
    }

    public int size() {
        return size;
    }

    public Object get(int index) {
        checkBounds(index);
        return array[index];
    }

    public ArrayList add(Object value) {
        expandIfNeeded();
        array[size()] = value;
        size++;
        return this;
    }

    public ArrayList prepend(Object value) {
        expandIfNeeded();
        shift();
        array[0] = value;
        size++;
        return this;
    }

    public Object delete(int index) {
        checkBounds(index);
        Object originalValue = array[index];
        shiftDown(index);
        size--;
        array[size()] = null;
        return originalValue;
    }

    public Object set(int index, Object value) {
        checkLowerBound(index);
        expandIfNeeded(index + 1);
        Object originalValue = array[index];
        array[index] = value;

        if (index >= size()) {
            size = index + 1;
        }

        return originalValue;
    }

    private void expandIfNeeded() {
        expandIfNeeded(size() + 1);
    }

    private void expandIfNeeded(int newSize) {
        if (newSize <= array.length) {
            return;
        }

        Object[] originalArray = array;
        int newLength = originalArray.length * 2;

        while (newLength < newSize) {
            newLength *= 2;
        }

        array = new Object[newLength];

        for (int i = 0; i < originalArray.length; i++) {
            array[i] = originalArray[i];
        }
    }

    private void shift() {
        for (int i = array.length - 1; i > 0; i--) {
            array[i] = array[i - 1];
        }
    }

    private void shiftDown(int index) {
        for (int i = index; i < array.length - 1; i++) {
            array[i] = array[i + 1];
        }
    }

    private void checkBounds(int index) {
        checkLowerBound(index);
        checkUpperBound(index);
    }

    private void checkLowerBound(int index) {
        if (index < 0) {
            throw new IndexOutOfBoundsException("Invalid index: " + index);
        }
    }

    private void checkUpperBound(int index) {
        if (index >= size()) {
            throw new IndexOutOfBoundsException("Invalid index: " + index);
        }
    }
}
